### PR TITLE
Convert `EnzoEFltArrayMap` to `ViewMap<T>` (& resolve issue #172)

### LINKED ISSUE #172: Unsound `const` semantics in `EnzoEFltArrayMap`
I recently realized that the `const` semantics of `EnzoEFltArrayMap` are unsound. 

The class was defined with the intention that `const` instances would only provide access to read-only data:
- When you access an element in a non-`const` instance of `EnzoEFltArrayMap`  (e.g. by invoking `array_map["density"]`), an instance of `CelloArray<enzo_float,3>` is returned. The elements of the returned array can freely modified.
- When you access an element in a `const` instance of `EnzoEFltArrayMap` (e.g. by invoking `array_map["density"]`), an instance of `CelloArray<const enzo_float,3>` is returned. In this case elements of the returned array can't be mutated.

Unfortunately, you can easily get around this intention by writing something like the following:
```c++
void func(const EnzoEFltArrayMap& array_map){
  // the function signature is meant to signal that the contents of array_map
  // don't get mutated by this function.
  EnzoEFltArrayMap array_map_copy = array_map;
  CelloArray<enzo_float,3> density = array_map_copy["density"];
  density(0,0,0) = 0.0;
  // because of the smart-pointer semantics of CelloArray, the above code mutates
  // the contents of array_map["density"]
}
```
The main point of this issue is to advise people to avoid writing code like the above. If you want to modify the contents of an `EnzoEFltArrayMap` don't declare it to be `const`.

This is similar to an issue I encountered with an early implementation `CelloArray`, which I solved by making `CelloArray<const T, 3>` indicate that contents could not be mutated. My inclination is to solve the problem in a similar way:
- move the implementation of `EnzoEFltArrayMap` to a new template class called `CelloArrayMap<T>`. All cases where `T` is a non-`const` type would provide access to mutable data, and all cases where `T` is a `const` type would only provide read-only access to the data. 
- define `EnzoEFltArrayMap` to alias `CelloArrayMap<enzo_float>`
- this is relatively straight-forward to do and would not break **any** existing code (even the "unsound" example code above). I'm just hesitant to introduce this change until some of my PRs start getting merged (to avoid merge conflicts).

As an aside: an alternative approach is to modify the signatures of copy/move constructors and assignment operations so that they can't accept `const` instances of `EnzoEFltArrayMap`. But, this may make the class more difficult to work with.

### PR DESCRIPTION
**Pull request by mabruzzo**
This PR migrates the contents of `EnzoEFltArrayMap`, which is a class encapsulating a map of `CelloView<enzo_float,3>` instances, to a new class template called `ViewMap<T>`. For backwards compatibility, `EnzoEFltArrayMap` is now a type-alias of  `ViewMap<T>`.

## Overview

In general, I think it makes a lot of sense for defining ``ViewMap`` alongside ``CelloView`` and it's something I've talked about doing for a long time. A major motivation for doing this is to resolve #172, which details some soundness issues with `EnzoEFltArrayMap`. The soundness issue and solution are described down below in the following spoiler tag.

<details> 
  <summary>Soundness problem & Solution </summary>

- Historically, we tried to define ``EnzoEFltArrayMap`` such that a ``const EnzoEFltArrayMap`` represented read-only data (we defined a few extra methods to help facilitate this). This can be useful when declaring function arguments.
- However, it later came to my attention that there was an easy way to mutate the data contained within a ``const EnzoEFltArrayMap`` that is passed to a function. The only way to fix is to disallow the creation of copies of ``EnzoEFltArrayMap`` (and to disable *move* semantics as well).
- To fix this I adopted a solution that is very similar to what we do in `CelloView` for this exact scenario. `ViewMap<T>` now represents the mutable datastructure while `ViewMap<const T>` represents the immutable data structure. I have defined a special constructor to facilitate implicit casts from `ViewMap<T>` to `ViewMap<const T>` that are about as seamless as the casts from `float*` to `const float*`.
- I have removed some methods that are now made unnecessary after making this change.
- I have added a section to the website-documentation discussing this behavior.

</details>

When reviewing this PR, I suggest that you step through the commits step-by-step (rather than just looking at the final files). I essentially completed the following steps:

1. I defined `ViewMap<T>` which is almost a perfect one-to-one copy of `EnzoEFltArrayMap` (a few minor changes were needed to make it work as a class template).
2. I wrote a number of unit tests for `ViewMap<T>` (to help do this, I factored out some functionality from `Cello/test_ViewCollec.cpp` into a new header file called `Cello/test_ViewTestRoutines.hpp`). These tests mostly check existing functionality that we already know works (since we've been using `EnzoEFltArrayMap` for a while).
3. I removed `EnzoEFltArrayMap` from the codebase and made `EnzoEFltArrayMap` into an alias for `ViewMap<enzo_float>`. I did not replace `EnzoEFltArrayMap` throughout the codebase because that would touch a LOT of files (we can revisit that in the future). I also updated the website-documentation accordingly.
4. I now refactored `ViewMap` to actually address the soundness problem (and I updated the website-documentation accordingly)
5. I updated the internal comments and variables within `ViewMap` so that they now discuss the contained views rather than the contained arrays (which is consistent with our previous renaming of `CelloArray` -> `CelloView`). I did NOT touch the names of methods because that would touch a LOT of additional files (we can also revisit this in the future).

## Question for reviewers?
Do you think it's worth renaming `ViewMap` to `View3DMap` to make it clear that it only holds 3D views (i.e. `CelloView<T,3>`)?

### CODE REVIEW

File: doc/source/devel/array.rst [hunk 1]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
+equivalent to ``scratch_view_map[key_l[i]]`` and
+``other_view_map[other_key_l[i]]``, in the context of the preceding
+snippets. Note that we don't support passing an integer value to
+``ViewMap::at``.
 
 Copy and ``const`` Semantics
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Making a copy of an ``EnzoEFltArrayMap`` instance (e.g. with a copy
+Making a copy of an ``ViewMap`` instance (e.g. with a copy
```
Comment (brenjohn) [inline doc/source/devel/array.rst:876]:
```suggestion
Making a copy of a ``ViewMap`` instance (e.g. with a copy
```

File: doc/source/devel/array.rst [hunk 2]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
-In comparison, element-access of a ``const EnzoEFltArrayMap`` yields a
-``CelloView<const enzo_float,3>`` which prevents direct modification of
-array elements.
+A ``const ViewMap`` holds very little meaning. If you declare and
+assign a variable, ``var``, as
+``const ViewMap<double> var = /*... */;``,
+then the fact that you used ``const ViewMap`` effectively tells the
+compiler to forbid you from performing ``var = other_view_map;`` in a
+subsequent statement.  You are still free to mutate the elements held
+of any contained CelloView.
```
Comment (brenjohn) [inline doc/source/devel/array.rst:889]:
```suggestion
by any contained CelloView.
```

File: doc/source/devel/array.rst [hunk 3]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
+
+If you want to denote that a ``ViewMap`` is read-only, you should
+declare declare the type as ``ViewMap<const double>`` instead. Note
```
Comment (brenjohn) [inline doc/source/devel/array.rst:892]:
```suggestion
declare the type as ``ViewMap<const double>`` instead. Note
```

File: doc/source/devel/array.rst [hunk 4]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
+that a ``ViewMap<float>`` can be implicitly converted to a
+``ViewMap<const float, N>`` (e.g. you can pass the former to a
```
Comment (brenjohn) [inline doc/source/devel/array.rst:894]:
```suggestion
``ViewMap<const float>`` (e.g. you can pass the former to a
```

File: doc/source/devel/array.rst [hunk 5]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
+    CelloView<enzo_float,3> calculate_pressure(ViewMap<const enzo_float> arg,
+                                               double density_floor);
+
+This function signature makes a very clear promise to other developers
+that ``arg`` is not mutated inside of the function. In fact the
+compiler will not compile any implementation of the function where the
+contents of arg are mutated (this can be a very useful reminder to any
+future developer that alters the implementation of
+``calculate_pressure`` - they are of course able to change the
+function signature, but it's a a good reminder to make sure that they
```
Comment (brenjohn) [inline doc/source/devel/array.rst:917]:
```suggestion
function signature, but it's a good reminder to make sure that they
```

File: doc/source/devel/array.rst [hunk 6]
```diff
@@ -853,75 +861,133 @@ or to replace the ``CelloView`` associated with a given
...
-     ``CelloView`` values in a single ``CelloView<enzo_float, 4>``
-     instance. Access of individual ``CelloView`` values is
-     accomplished with the overload of the
-     ``CelloView<T,D>::subarray`` method. This approach is used when
-     you construct an ``EnzoEFltArrayMap`` that allocates memory for
-     the contained ``CelloView``\s.
+     ``CelloView`` values in a single ``CelloView<T, 4>`` instance.
+     Access of individual ``CelloView`` values is accomplished with
+     the overload of the ``CelloView<T,D>::subarray`` method. This
+     approach is used when you construct an ``ViewMap`` that
```
Comment (brenjohn) [inline doc/source/devel/array.rst:989]:
```suggestion
     approach is used when you construct a ``ViewMap`` that
```

File: src/Cello/view_ViewMap.hpp [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+  /// key-order matches expectations.
+  ///
+  /// @param ref The list of reference keys
+  /// @param raise_err When true, this will raise a fatal error if the order
+  ///    doesn't match
+  /// @param allow_smaller_ref When `true`, the key-order is only compared for
+  ///    first `ref.size()`. When `false` (the default), the map must have the
+  ///    same number of keys as `ref`.
+  bool validate_key_order(const std::vector<std::string> &ref,
+			  bool raise_err, bool allow_smaller_ref = false)
```
Comment (brenjohn) [inline src/Cello/view_ViewMap.hpp:179]: Maybe it makes sense to have a default value for `raise_err` here also?

File: src/Cello/test_ViewMap.cpp [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+    // getting the signatures on the template functions correct so that they
+    // can be used to compare the template functions is challenging! So we just
+    // do it manually!
+    bool all_equal = true;
+    for (const std::string& key: src_keys) {
+      CelloView<const int,3> dest_view = dest_view_map.at(key);
+      CelloView<const int,3> src_view = src_view_map.at(key);
+
+      const int mz = dest_view.shape(0);
+      const int my = src_view.shape(1);
```
Comment (brenjohn) [inline src/Cello/test_ViewMap.cpp:685]:
```suggestion
      const int my = dest_view.shape(1);
```
I guess this really doesn't make a difference but src_view is a little unexpected when reading this.

**Review by brenjohn — APPROVED** [review]
I think this looks great as it is. I only had very minor comments/suggestions.

**Comment by brenjohn** [comment]
As for renaming `ViewMap` to `View3DMap`, I think that makes sense but there was a suggestion in the documentation to generalize the implementation which would make a 2D version redundant. Maybe it makes sense to add a note in the documentation that this only works for 3D views at the moment and then we can look at generalizing to 2D views when someone needs them?

File: src/Cello/test_ViewTestRoutines.hpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     test_ViewTestRoutines.cpp
+/// @author   Matthew Abruzzo (matthewabruzzo@gmail.com)
+/// @date     2023-07-03
+/// @brief    Common routines used for testing CelloView and related machinery
+
+#ifndef TEST_VIEW_TEST_ROUNTINES_HPP
```
Comment (bwoshea) [inline src/Cello/test_ViewTestRoutines.hpp:8]: I'm not entirely sure what a "ROUNTINE" is... I think that this is a typo (same note for the line immediately below)

**Review by bwoshea — APPROVED** [review]
Outside of one minor comment I don't have much to add beyond what @brenjohn had to say.  It looks good to me!

**Comment by bwoshea** [comment]
@mabruzzo it looks like there are some minor suggestions that should be very easy to resolve, plus merge conflicts that need to be fixed.  If you can do those we can merge this PR!

File: src/Cello/view_ViewMap.hpp (hunk 7 above)
Comment (mabruzzo) [reply src/Cello/view_ViewMap.hpp:179]: added

File: src/Cello/test_ViewMap.cpp (hunk 8 above)
Comment (mabruzzo) [reply src/Cello/test_ViewMap.cpp:685]: Good point! fixed

File: src/Cello/test_ViewTestRoutines.hpp (hunk 9 above)
Comment (mabruzzo) [reply src/Cello/test_ViewTestRoutines.hpp:8]: Thanks for catching that!

**Comment by mabruzzo** [comment]
I know the answer-tests started failing before for commit [0d99309](https://github.com/enzo-project/enzo-e/pull/326/commits/0d9930996d4973d12996d370820bbd5371468e47).

But, I think that may be related to a grackle-change and the fact that we don't pin our dependencies entirely correctly. A little while back, we had to update the gold-standard to deal with this issue. I think that commit [0d99309](https://github.com/enzo-project/enzo-e/pull/326/commits/0d9930996d4973d12996d370820bbd5371468e47) didn't include those changes.

If I'm right about this, then merging in the main branch just now should cause the tests to pass.

**Pull request merged**